typing/jkind.ml

```diff
@@ -356,56 +363,332 @@ module Error = struct
+          else Stop
+        | Tconstr (p, args, _) -> (
+          match Path.Map.find_opt p constr with
+          | None ->
+            Continue
+              { t with
+                constr = Path.Map.add p (initial_fuel_per_ty, args) constr
+              }
+          | Some (fuel, seen_args) ->
+            if List.for_all2 type_equal seen_args args
+            then Skip
+            else if fuel > 0
+            then
+              Continue
+                { t with constr = Path.Map.add p (fuel - 1, args) constr }
+            else Stop)
+        | Tvar _ | Tarrow _ | Tunboxed_tuple _ | Tobject _ | Tfield _ | Tnil
+        | Tvariant _ | Tunivar _ | Tpackage _ ->
+          (* these cases either cannot be infinitely recursive or their jkinds
+             do not have baggage *)
+          Continue t
+        | Tlink _ | Tsubst _ ->
+          Misc.fatal_error "Tlink or Tsubst in reduce_baggage"
+    end in
+    let (module A) = Axis.get axis in
+    let rec loop ctl bound_so_far = function
+      (* early cutoff *)
+      | _ when A.le A.max bound_so_far -> bound_so_far
+      | [] -> bound_so_far
+      | b :: bs -> (
+        match Loop_control.check ctl b with
+        | Stop -> A.max (* out of fuel *)
+        | Skip -> loop ctl bound_so_far bs (* skip [b] *)
+        | Continue ctl_after_unpacking_b -> (
+          match jkind_of_type b with
+          | Some b_jkind ->
+            let b_bound = Bounds.get ~axis b_jkind.jkind.upper_bounds in
+            let bound_so_far = A.join bound_so_far b_bound.modifier in
+            let bound_so_far =
+              loop ctl_after_unpacking_b bound_so_far
+                (Baggage.as_list b_bound.baggage)
+            in
+            (* Use *original* ctl here, so we don't fall over on
+               a record with 20 lists with different payloads. *)
+            loop ctl bound_so_far bs
+          | None ->
+            (* hd is not principally known, so we treat it as having the max
+               bound *)
+            (* CR layouts v2.8: Does this ever trigger? Richard is skeptical
+               that we need to worry about principality here. *)
+            A.max))
+    in
+    loop Loop_control.starting modifier baggage
+
+  let reduce ~type_equal ~jkind_of_type ~axis bound =
+    reduce_baggage ~type_equal ~jkind_of_type ~axis bound.modifier
+      (Baggage.as_list bound.baggage)
+
+  let less_or_equal :
+      type axis l r.
+      type_equal:_ ->
+      jkind_of_type:_ ->
+      axis:axis Axis.t ->
+      (allowed * r, axis) t ->
+      (l * allowed, axis) t ->
+      Misc.Le_result.t =
+   fun ~type_equal ~jkind_of_type ~axis { modifier = m1; baggage = b1 }
+       { modifier = m2; baggage = b2 } ->
+    let (module Axis_ops) = Axis.get axis in
+    match b1, b2 with
+    | No_baggage, No_baggage -> Axis_ops.less_or_equal m1 m2
+    (* CR layouts v2.8: This should expand types on the left. *)
+    | Baggage (ty, tys), No_baggage ->
+      if Axis_ops.le Axis_ops.max m2
```

Comment:
This test is redundant since `reduce_baggage` terminates early when the bound is max